Chain dma support

jsarha opened the pull request:
Its probably worth squashing this up a bit, especially now that I had to mangle Ranjani's commit quite a bit when solving sof-dev rebase conflicts.  But maybe this is easier to review in pieces.

Both 16- and 32-bit audio works with this PR, but getting 44.1kHz need still some more work.

RanderWang commented:
Could you give a full picture of solution ?  How to use chain dma ?   How to support in topology ?  How does driver do to cooperate with topology?

ranj063 commented:
@jsarha can you please squash the relevant patches and add descriptions? It is very hard to review it like this.

Also, IPC4 test fails spectacularly. So something is not right.

jsarha commented:
Here is the squashed up version, but there is still something severely wrong with normal playback, and I am in the middle of debugging it.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+				return 0;
+		}
+	}
+
```
can we move this to a separate function, this is becoming really hard to follow.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+			}
+
+			/* reset is preceded by pause */
+			if (state == SOF_IPC4_PIPE_RESET)
```
I would use a switch(state) and trap the default as an error. It's not really possible based on states implemented in sof_ipc4_pcm_trigger but it would be more consistent.

jsarha replied:
So it appear from the prints I've added, but the inner workings of the drivers IPC4 support are not that clear to me yet. But if you say that anything but SOF_IPC4_PIPE_RUNNING, SOF_IPC4_PIPE_PAUSED, and SOF_IPC4_PIPE_RESET is an error, I'll do the change.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+		swidget = widget->dobj.private;
+
+		if (!swidget)
+                        continue;
```
can this happen, and if yes is this a feature or a bug?

jsarha replied:
Not sure. This is part of my pedantic copying of the behavior before rebasing on top of latest topic/sof-dev, so  I copied the old loop in every detail from previous branch. In the end the problem was not in this function at all. In the next version this loop is completely rewriten.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+				msg->primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+				msg->primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
+				msg->primary |= SOF_IPC4_GLB_CHAIN_DMA_ALLOCATE_MASK;
+				msg->primary |= SOF_IPC4_GLB_CHAIN_DMA_ENABLE_MASK;
```
probably need comments to explain why we have to do allocate and enabled at the same time

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1116,6 +1132,27 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			u32 fifo_size;
+
+			host_dma_id = platform_params->stream_tag - 1;
+			pipeline->msg.primary &= ~SOF_IPC4_GLB_CHAIN_DMA_HOST_ID_MASK;
```
this is already cleared on line 973

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1116,6 +1132,27 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+				pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
+
+			fifo_size = 2 * ((params_rate(fe_params) * params_channels(fe_params) *
+					  params_width(fe_params))/ 8000);
```
that formula looks like art or black magic. What is the '2' factor for and why would we divide by 8000?

jsarha replied:
Oh, this came from @ranj063, and I only added the sample width to the equation. It gives reasonable nubers for 48kHz stereo 16 and 32-bit (384 and 768 bytes), but now that I look the corresponding fw code I am not sure this would even work for 44.1kHz:
https://github.com/thesofproject/sof/blob/eb712a40db3dae1be73acda772b2e62dd73a92a1/src/ipc/ipc4/helper.c#L499

But what would be better? The size of required buffer should anyway be proportional to sample-rate, sample-width, and number channels, no?

ranj063 replied:
> that formula looks like art or black magic. What is the '2' factor for and why would we divide by 8000?

8 is for converting to bytes and 1000 to converto to ms. So the fifo_size should be a minimum of 2ms buffer.

plbossart replied:
shouldn't we use snd_pcm_format_physical_width, i.e. the container size, rather the number of bits valid?

I find it scary that we'd hard-code the tick size. It might be better to have a define or a topology configuration that we make sure if larger than 2ms.

also I don't get why it's exposed as a fifo, this is a DMA buffer isn't it?

jsarha replied:
'fifo_size' is the name used in cSOF Firmware Interface spec.

plbossart replied:
that's worthy of a comment, there's another FIFO in HDaudio but a real-one :-)

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -292,16 +296,29 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+	 *       audio, which is playback only. If chain dma is ever
+	 *       used for recording, then extracting the pipeline from
+	 *       rtd may need some more lines.
+	 */
```
we should absolutely try the capture with the HDaudio external codec. This would be a really good tool to debug.

We should even have a mode where we override the topology and force the pipelines to be dma-chained with a bit set in sof_debug.

jsarha replied:
Now that I think about it, the pcms are symmetrical, no? So in such a case the playback substream should have the use_chain_dma property too, shouldn't it? Or should I check if either direction has the property and get rid of the mask in both cases? Or do we need something more complex?

plbossart reviewed (commented):
This is encouraging @jsarha, this PR needs more comments, background information and needs a bit of polishing.

I think we should really try the capture as well, this could be really really useful to hardware validation. In the end we would have 
a) DSP-less
b) chained DMA
c) full firmware support with decoupled DMAs.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -66,16 +66,66 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 		return -EINVAL;
 
+        list = spcm->stream[substream->stream].list;
+        for_each_dapm_widgets(list, num_widgets, widget) {
```
@jsarha this step is incorrect. Why would you loop through the widgets here and down again in line 130 onwards. What you need is to modify the loop here:
```
for (i = pipeline_list->count - 1; i >= 0; i--) {
		pipe_widget = pipeline_list->pipe_widgets[i];
		pipeline = pipe_widget->private;
		if (pipeline->state != state && !pipeline->skip_during_fe_trigger)
			data->pipeline_ids[data->count++] = pipe_widget->instance_id;
	}
```

to do something like 
```
for (i = pipeline_list->count - 1; i >= 0; i--) {
		pipe_widget = pipeline_list->pipe_widgets[i];
		pipeline = pipe_widget->private;
		if (pipeline->use_chain_dma)
                               return sof_ipc4_chaind_dma(sdev, pipe_widget);
	}
```

jsarha replied:
Yes, I had it implemented differently on my first rebase attempt, but since the first attempt did not work, on the second round I tried to follow the original implementation more closely, but in the end the problem was not here. I'll move the implementation to this loop again, just because its less code that way, but are you sure the current implementation is technically wrong? Is there some problem that I do not see here?

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -284,6 +284,10 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+	struct snd_soc_dai *cpu_dai = asoc_rtd_to_cpu(rtd, 0);
+	struct snd_soc_dapm_widget *w;
+	struct snd_sof_widget *swidget, *pipe_widget;
+	struct sof_ipc4_pipeline *pipeline;
```
The commit message got a bit mixed up, can you re-phrase it?

and re-order the variables.

jsarha commented:
All the issues found in the review are not yet addressed, but pushed the branch again to trigger CI tests to see if the nocodec problems persist.

jsarha commented:
I don't understand why details links lead to 404. Maybe another push will help. There is some more fixes this time too. I think this is the last unresolved issue: https://github.com/thesofproject/linux/pull/3953#discussion_r1005856837

ranj063 commented:
> I don't understand why details links lead to 404. Maybe another push will help. There is some more fixes this time too. I think this is the last unresolved issue: [#3953 (comment)](https://github.com/thesofproject/linux/pull/3953#discussion_r1005856837)

01.org is down. Can you check the internal link @jsarha

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -192,6 +192,35 @@ enum sof_ipc4_pipeline_state {
+#define SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE(x)	   (((x) << \
+						     SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE_SHIFT) & \
+						    SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE_MASK)
+
```
ASoC: SOF: IPC4: chain-dma: Add macros for chain-dma message bits

plbossart replied:
also need an explanation of what 'Chain DMA' is

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -475,36 +475,43 @@ static int ipc4_hda_dai_trigger(struct snd_pcm_substream *substream,
+							  SOF_IPC4_PIPE_PAUSED);
+			if (ret < 0)
+				return ret;
+		}
```
@ranj063 this is the sort of things that should be abstracted, this will also be the same in DSP-less mode.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -59,6 +59,56 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		/* Set allocate and enable bits to configure and enable chain DMA. */
+		return sof_ipc4_chain_dma_send_msg(sdev, msg, true, true);
+	case SOF_IPC4_PIPE_PAUSED:
+		/* Set allocate bit, but not enable bit. */
```
I think it's enough to say "disable chained DMA"

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -59,6 +59,56 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		/* Set allocate bit, but not enable bit. */
+		return sof_ipc4_chain_dma_send_msg(sdev, msg, true, false);
+	case SOF_IPC4_PIPE_RESET:
+		/* Set allocat and ennabel bits to zero. */
```
and "disable and free chained DMA" here

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -59,6 +59,56 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		/* Set allocat and ennabel bits to zero. */
+		return sof_ipc4_chain_dma_send_msg(sdev, msg, false, false);
+	default:
+		dev_err(sdev->dev, "%s: unexpected state %d", __func__, state);
```
no need for %s __func__

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -82,6 +132,14 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	if (!pipeline_list->pipe_widgets)
 		return 0;
 
+	/* If this is a chained dma stream, then there is no need to continue further. */
```
to be more explicit, we should say something like "Streams using chained DMA contain a single pipeline, so trigger that pipeline and return"

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1116,6 +1126,30 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			pipeline->msg.primary &= ~SOF_IPC4_GLB_CHAIN_DMA_HOST_ID_MASK;
+			pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(host_dma_id);
+
+			if (params_width(fe_params) > 16)
```
a comment here saying "Set SCS bit for all formats other than S16_LE"

jsarha replied:
Then probably better to change the condition to exactly that too.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1116,6 +1126,30 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			else
+				pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
+			/*
+			 * Despite its name the bitfield 'fifo_size' is used to define dma buffer
```
DMA in caps

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1116,6 +1126,30 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			 */
+			fifo_size = 2 * ((params_rate(fe_params) * params_channels(fe_params) *
+					  params_physical_width(fe_params))/ 8000);
+			dev_dbg(sdev->dev, "fifo_size is %d\n", fifo_size);
```
can we remove the debug here?

jsarha replied:
Sure.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1145,6 +1179,12 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+		struct snd_sof_widget *pipe_widget = swidget->pipe_widget;
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+		if (pipeline->use_chain_dma)
```
I think we should add a check here to make sure that only HDA DAI's are using the chained DMA. In fact prepare() might be a bit too late, we should do this in sof_ipc4_widget_setup_comp_dai(). The tricky part will be to make sure that the pipeline widgets are parsed first. I have a feeling they are but best to check.

ranj063 replied:
BTW are you missing something in the unprepare_copier_module()

jsarha replied:
I dropped them on purpose. I found it clearer and more fool proof (me being the fool not being certain in which order the functions are called), to zero the dma id bits right before assigning new values there, like I do with scs bit and the fifo_size. Felt unsymmetrical doing the zeroing differently with different bits. But if you think its better to go back to old and maybe even drop the zeroing before the assignment in dma id cases, then sure I'll go back.

ranj063 replied:
I think it would be more balanced to set the bits in prepare() and clear them in unprepare()

jsarha replied:
> I think we should add a check here to make sure that only HDA DAI's are using the chained DMA. In fact prepare() might be a bit too late, we should do this in sof_ipc4_widget_setup_comp_dai(). The tricky part will be to make sure that the pipeline widgets are parsed first. I have a feeling they are but best to check.

@ranj063 I've addressed all the other comments, but this is a bit tricky as in sof_ipc4_widget_setup_comp_dai()  the associated pipeline object is not yet built, so I do not know any easy way to dig out the use_chain_dma property. So I assume I should do the check in sof_ipc4_prepare_copier_module() after all, or is there something I overlooked?

jsarha replied:
So I added the check in sof_ipc4_prepare_copier_module().

ranj063 replied:
@jsarha there's likely 2 things you have to do to be able to do the check in sof_ipc4_widget_setup_comp_dai(). Set the scheduler for all widgets before we set up the IPC structures and likely also traverse the list in reverse. Let me send you a patch for this a little later

ranj063 reviewed (commented):
Looking a lot better @jsarha. A few more comments from me

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2120,23 +2120,7 @@ static int sof_complete(struct snd_soc_component *scomp)
 			}
 		}
 
-	/*
```
@jsarha can you please make this the first patch?

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -492,6 +494,16 @@ static int sof_ipc4_widget_setup_comp_dai(struct snd_sof_widget *swidget)
 
 	ipc4_copier->data.gtw_cfg.node_id = SOF_IPC4_NODE_TYPE(node_type);
 
+	pipe_widget = swidget->pipe_widget;
```
and squash this with the second patch?

jsarha replied:
Done.

jsarha commented:
Hmm, now that I tested this latest change a  bit, it does not appear to work as it should. hw:0,0 does not work any more.

jsarha commented:
The first patch was updated and now this appears to work Ok.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -234,24 +292,43 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 		return -EINVAL;
 	}
 
+	for_each_pcm_streams(dir) {
```
nit-pick. can we add a comment saying we should not change format when using chained DMA

jsarha replied:
The comment is there couple of lines down, where the format mask is set, but maybe it could be more explicit.

jsarha replied:
Oh, and there is no reason to check the use_chain_dma flag earlier than just before setting the mask.

ranj063 reviewed (approved):
looks good @jsarha . just a minor comment but that can be a fixup

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2150,12 +2144,29 @@ static int sof_complete(struct snd_soc_component *scomp)
 					if (ret < 0)
 						return ret;
 				}
+
```
spurious newline?

ranj063 replied:
no, there's a variable declaration just before right?

plbossart replied:
the code was modified, the comment is irrelevant now

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2150,12 +2144,29 @@ static int sof_complete(struct snd_soc_component *scomp)
 					if (ret < 0)
 						return ret;
 				}
+
```
in the loop above, should we skip the case where swidget->id == comp_swidget->id?

It's a bit confusing when some widgets belong to another type of widget....

jsarha replied:
I think this issue got fixed in the latest rebase.

ranj063 replied:
not following @plbossart . the first loop is for the pipeline_list and it contains only scheduler widgets

plbossart replied:
I don't understand what GitHub is doing, the comment was about a double loop which has now vanished.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -192,6 +192,35 @@ enum sof_ipc4_pipeline_state {
 #define SOF_IPC4_GLB_LOAD_LIBRARY_LIB_ID_SHIFT	16
 #define SOF_IPC4_GLB_LOAD_LIBRARY_LIB_ID(x)	((x) << SOF_IPC4_GLB_LOAD_LIBRARY_LIB_ID_SHIFT)
 
+/* chain dma ipc message */
```
suggested edit for the commit message

In the chained DMA mode, the firmware allocates buffers for the host and link DMA, and takes care of copying data between host- and link-DMA buffers in a low-latency thread. This is different to a regular pipeline, no processing is allowed, and the connection between host- and link DMA is handled with a dedicated IPC.

This patch exposes the macros needed to create the required IPC messages.

jsarha replied:
Thanks! Writing a good description for a commit is often more difficult than writing the code for it. Especially when one do not yet know for certain all the relevant details 😅.

jsarha commented:
Addressed the last comment, but there is conflicts now... rebasing.

plbossart commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -54,6 +54,7 @@
 #define SOF_TKN_SCHED_DYNAMIC_PIPELINE		206
 #define SOF_TKN_SCHED_LP_MODE			207
 #define SOF_TKN_SCHED_MEM_USAGE			208
+#define SOF_TKN_SCHED_USE_CHAIN_DMA		209
```
The commit message needs to be rewritten, too verbose and unclear at the same time

suggested edits:

Add logic for setting up and tearing down chained DMA connections. 

Since pipelines are not used, all the logic to set the pipeline states can be bypassed, with only the DMA programming sequences remaining. In addition the same format needs to be used for host- and link-DMA, without the usual fixup to use the S32_LE format on the link.

Note however that for convenience and compatibility with existing definitions, the topology relies on the concept of pipelines with a 'USE_CHAIN_DMA' token indicating that all the logic shall be bypassed.

Unlike 'normal' ALSA sequences, the chain DMA is not programmed in hw_params/hw_free. The IPC message tp set-up and tear-down chained DMA are sent in sof_ipc4_trigger_pipelines(), but the contents prepared earlier.

Chained DMA is only supported by the Intel HDA DAI for now.

jsarha replied:
Thanks! Certainly better than my version that has evolved over time, and where I've avoided subjects that I am not certain of. With your permission, I take that into use.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -249,9 +311,24 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+	}
+
+	/* Always set BE format to 32-bits for both playback and capture, except if chained
+	 * DMA is in use. In chained DMA case the back-end and front-end format must match. */
```
which begs the question: how is this match done and how are the codec_dai restrictions taken into account?

This is a similar problem to the dspless mode added by @ujfalusi

ranj063 replied:
@plbossart could you please spell out the codec DAI restriction we need to handle with chained DMA?

plbossart replied:
@ranj063 @jsarha  see https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-dai.c#L240

````c
	p_params.ch = params_channels(params);
	p_params.s_freq = params_rate(params);
	p_params.link_index = hlink->index;
	p_params.format = params_format(params);

	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
		p_params.link_bps = codec_dai->driver->playback.sig_bits;
	else
		p_params.link_bps = codec_dai->driver->capture.sig_bits;
````		

if the firmware does not modify the format, then don't wee need to change the format on the host side so that it matches what the codec can do?

ranj063 replied:
Ahh so we need to something similar to what @ujfalusi did as well ie save the format_val in hext_stream right?

plbossart replied:
possibly, I am not sure what the firmware does and if this is a problem that needs solving, but the parallel with @ujfalusi 's work is interesting.

jsarha replied:
I need a little digesting for this. ATM according to my tests everything else works OK with chained dma hdmi cpm, but only 48kHz is supported, and the PCM claims to support S24_LE format, but it does not.  The S24_LE samples are severely attenuated, probably due to the S24 being interpreted as S32 and 8msb bits being zero. Anyway I push the branch as it is now without addressing this at least to see if its otherwise good now.

ranj063 replied:
@jsarha can we add a comment in the commit message saying that only 16/32 bits are supported with chained DMA for now?

RanderWang replied:
so we need to change it in topology that only 16bit/32bit are support. FW doesn't change format so 24bit is a problem.

plbossart reviewed (changes requested):
Mostly editorial review. The only open I have is how the codec DAI limitation is handled, if at all?

jsarha commented:
Yesterdays version now pushed again with @plbossart 's suggestions for descriptions and after a rather painful rebase. I am currently looking at how to add codec side format constraints to a chained dma pcm.

ranj063 commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -97,7 +98,6 @@
 #define SOF_TKN_COMP_SINK_PIN_BINDING_WNAME	413
 #define SOF_TKN_COMP_SRC_PIN_BINDING_WNAME	414
 
-
```
spurious change

jsarha replied:
Looks like 9331aa0feb774 fixed this for me.

jsarha commented:
Updated to work with this latest sof topology pr: https://github.com/thesofproject/sof/pull/6524

bardliao commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,60 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+		struct snd_sof_pipeline *spipe = pipeline_list->pipelines[i];
+		struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
```
Should we add
```
if (!pipeline->use_chain_dma)
        continue;
```
here?

jsarha replied:
I do not think we should be in this function with a pipeline_list that has any pipeline without pipeline->use_chain_dma. Maybe the check with an error print and return -EINVAL? Or what do you think @ranj063 ?

ranj063 replied:
that check would be redundant and will always be false. No need to add the check

jsarha replied:
Well, I guess we can not check everything all the time.

bardliao replied:
@ranj063 @jsarha But see line 279 to 288
```
	for (i = pipeline_list->count - 1; i >= 0; i--) {
		spipe = pipeline_list->pipelines[i];
		pipe_widget = spipe->pipe_widget;
		pipeline = pipe_widget->private;

		/* If a pipeline in the list uses chained DMA, trigger chained DMA and return */
		if (pipeline->use_chain_dma)
			return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state);
	}
```
We will call `sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state);` if there is one pipeline that use_chain_dma is true.
If I understand it correctly, other pipelines in the pipeline_list could be use_chain_dma = false. Am I correct?
How can we guaranty that all pipelines in the pipeline_list  are all use_chain_dma = true?

jsarha replied:
The quoted loop does not run to the end, so there is really only one check. But yes, I can see now why this is confusing. I'll add a comment here too. In fact even the loop is unnecessary. Checking use_chain_dma from the first element in the pipelist would be enough. However, to make sure the tplg is correct, I keep the loop and then add a check in sof_ipc4_chain_dma_trigger() that all elements in list have the attribute, since that is a requirement for the feature to work.

bardliao replied:
> The quoted loop does not run to the end, so there is really only one check. But yes, I can see now why this is confusing. I'll add a comment here too. In fact even the loop is unnecessary. Checking use_chain_dma from the first element in the pipelist would be enough. However, to make sure the tplg is correct, I keep the loop and then add a check in sof_ipc4_chain_dma_trigger() that all elements in list have the attribute, since that is a requirement for the feature to work.

Thanks @jsarha for the explanation, I understand it now.

bardliao commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1195,6 +1226,27 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+
+			/* Set SCS bit for S16_LE format only */
+			if (params_format(fe_params) == SNDRV_PCM_FORMAT_S16_LE)
+				pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_SCS_MASK;
```
`msg.primary |= pipeline->msg.primary;` is used in sof_ipc4_chain_dma_send_msg. What if some pipelines are S16_LE format and others are S32_LE format?

jsarha replied:
In chained dma case there there should be only two pipelines. One at host dai and another at hda end. Since there is no processing involved, the format must be the same for the pipelines.

ranj063 replied:
@bardliao pipelines exist only for the sake of tplg-driver compatibility. We do no use them in the DSP at all so the pipeline settings are meaningless for chained DMA.

RanderWang commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1973,9 +2057,17 @@ static int sof_ipc4_route_free(struct snd_sof_dev *sdev, struct snd_sof_route *s
 
+	/* no route is set up if chain DMA is used */
+	if (src_pipeline->use_chain_dma || sink_pipeline->use_chain_dma)
+		return 0;
```
will we have two pipelines for chain dma ?

ranj063 replied:
@RanderWang the tplg will have 2 pipelines yes. But the pipelines themselves have no meaning except to keep the tplg same for chain DMA and regular PCM

RanderWang commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1040,6 +1068,9 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
 			struct sof_ipc4_alh_configuration_blob *blob;
 			unsigned int group_id;
 
+			if (pipeline->use_chain_dma)
```
do we need to use chain dma for "ipc4_copier->dai_type == SOF_DAI_INTEL_ALH"

ranj063 replied:
no we dont. ALH uses the GPDMA

jsarha replied:
Oh, so this "if (pipeline->use_chain_dma)" is actually badly placed in side "if (ipc4_copier->dai_type == SOF_DAI_INTEL_ALH)". I'll move it outside.

RanderWang commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1031,6 +1053,12 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
 
 	if (WIDGET_IS_AIF(swidget->id)) {
+		if (pipeline->use_chain_dma) {
+			pipeline->msg.primary &= ~SOF_IPC4_GLB_CHAIN_DMA_HOST_ID_MASK;
```
why not pipeline->msg.primary = 0;  pipeline->msg.extension = 0; ? do we need to save any bits ?

jsarha replied:
I was tempted to do that before, but did not because it leads to double zeroing of link dma id bits, when we earlier had only one pipeline for both the host and hda end. However, now that there is two pipelines and two msg -structs it makes sense to zero both in their appropriate places.

RanderWang commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,60 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+	 * Copy the Link DMA ID, host DMA ID, fifo_size and the SCS bit settings from the pipelines.
+	 */
+	for (i = 0; i < pipeline_list->count; i++) {
+		struct snd_sof_pipeline *spipe = pipeline_list->pipelines[i];
```
more than 1 pipeline ? msg.primary and msg.extension will be set multiple times

ranj063 replied:
@RanderWang this is fully intentional. primary has teh host DMA ID and link DMA ID and they come from different pipelines. The extension contains the fifo_size and only comes from the FE pipeline. The BE pipeline never sets it. So it will always be 0. @jsarha maybe capture it with a comment?

jsarha commented:
> Mostly editorial review. The only open I have is how the codec DAI limitation is handled, if at all?

Since this feature is enabled through topology, would it be enough to code the codec dai limitations there too? I think I could dig the hda dai limitations and apply them as constraints to sof pcm when chained dma is in use, but is that really necessary?

plbossart commented:
> > Mostly editorial review. The only open I have is how the codec DAI limitation is handled, if at all?
> 
> Since this feature is enabled through topology, would it be enough to code the codec dai limitations there too? I think I could dig the hda dai limitations and apply them as constraints to sof pcm when chained dma is in use, but is that really necessary?

In general anything tied to hardware codec support cannot be part of the topology.

That said you bring-up an interesting point @jsarha. It turns out that the only codec DAI limitation is the 'sig_bits', but those are fixed in sound/soc/codecs/hdac_hda.c. I thought there were determined based on codec capabilities. @kv2019i do you think we could simplify all this, or get the values from the codec directly?

````c
static struct snd_soc_dai_driver hdac_hda_dais[] = {
{
	.id = HDAC_ANALOG_DAI_ID,
	.name = "Analog Codec DAI",
	.ops = &hdac_hda_dai_ops,
	.playback = {
		.stream_name	= "Analog Codec Playback",
		.channels_min	= 1,
		.channels_max	= 16,
		.rates		= SNDRV_PCM_RATE_8000_192000,
		.formats	= STUB_FORMATS,
		.sig_bits	= 24,
	},
	.capture = {
		.stream_name    = "Analog Codec Capture",
		.channels_min   = 1,
		.channels_max   = 16,
		.rates = SNDRV_PCM_RATE_8000_192000,
		.formats = STUB_FORMATS,
		.sig_bits = 24,
	},
},
````

Adding @ranj063 and @ujfalusi since we have the same codec dai issue for the abstraction and dsp-less PRs.

Edit: in the AVS driver, the DAI information is created dynamically based on codec information, see 
https://github.com/thesofproject/linux/blob/551c11c310acbe5a6636a1eecedc5c762f407222/sound/soc/codecs/hda.c#L55

````c		
                stream->channels_min = pcm->stream[dir].channels_min;
		stream->channels_max = pcm->stream[dir].channels_max;
		stream->rates = pcm->stream[dir].rates;
		stream->formats = pcm->stream[dir].formats;
		stream->sig_bits = pcm->stream[dir].maxbps;

````

RanderWang commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1195,6 +1226,27 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			u32 host_dma_id;
+			u32 fifo_size;
+
+			host_dma_id = platform_params->stream_tag - 1;
```
please add 

copier_data->gtw_cfg.node_id = 0xffffffff //invalid node id

jsarha commented:
On more round of applied review fixes, while waiting for https://github.com/thesofproject/linux/pull/3972 to get ready. The sig_bits are still not copied from codec side, but that should not make big functional difference. I'll see if I get it done before #3972 and this is merged , if not I'll do it as a separate PR.

jsarha commented:
Latest changes: Make sure the tplg is using use_chain_dma correctly. I keep the loop in sof_ipc4_trigger_pipelines() and then add a check in sof_ipc4_chain_dma_trigger() that all elements in list have the attribute, since that is a requirement for the feature to work.  Also improve comments in both places.

RanderWang commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1215,13 +1250,39 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			pipeline->msg.extension |= SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE(fifo_size);
+
+			/* Set node_id to invalid value. It shouldn't be used with chained DMA. */
+			copier_data->gtw_cfg.node_id = 0xffffffff;
+
+			return 0;
```
use SOF_IPC4_INVALID_NODE_ID.  thanks!

jsarha replied:
Done, thanks!

jsarha commented:
Only the 0xffffffff => SOF_IPC4_INVALID_NODE_ID change.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,70 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+	 * There are two pipelines involved, where the data is collected. The link side pipeline and
+	 * the host side pipeline. Part of the data needed is collected from the link side to the
+	 * link side pipeline->msg and part from the host side to host side pipeline->msg. The loop
+	 * bellow collects the data to a single IPC message.
```
typo: below

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,70 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+	/*
+	 * Copy the Link DMA ID, host DMA ID, fifo_size and the SCS bit settings from the pipelines.
+	 * There are two pipelines involved, where the data is collected. The link side pipeline and
+	 * the host side pipeline. Part of the data needed is collected from the link side to the
```
The link side pipeline and the host side pipeline.

Missing verb/complement or this needs to be part of the previous sentence. 

"part of the data" used in two different contexts sounds odd

Consider rewording, e.g.

To set-up the DMA chain, the host DMA ID and SCS setting are retrieved from the host pipeline configuration. Likewise the  link DMA ID and fifo_size are retrieved from the link pipeline configuration.

jsarha replied:
Must have been really tired when writing this piece of comment. I take you version, thanks.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -216,6 +282,21 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 
+	/*
+	 * Every pipeline of a chained DMA stream should have use_chain_dma attribute, so
+	 * this loop is only necessary for checking the correctnes of the topology. The loop
```
typo: correctness

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -383,9 +468,24 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
+	}
+
+	/* Always set BE format to 32-bits for both playback and capture, except if chained
+	 * DMA is in use. In chained DMA case the back-end and front-end format must match. */
```
````
WARNING: Block comments use a trailing */ on a separate line
#264: FILE: sound/soc/sof/ipc4-pcm.c:484:
+	 * DMA is in use. In chained DMA case the back-end and front-end format must match. */
````

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1215,13 +1250,39 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			 * size. The expression calculates 2ms buffer size.
+			 */
+			fifo_size = 2 * ((params_rate(fe_params) * params_channels(fe_params) *
+					  params_physical_width(fe_params))/ 8000);
```
missing space before / 

````
CHECK: spaces preferred around that '/' (ctx:VxW)
#383: FILE: sound/soc/sof/ipc4-topology.c:1277:
+					  params_physical_width(fe_params))/ 8000);
 					                                   ^
````

plbossart replied:
Should there be a ROUND_UP in case we ever support 44.1 kHz?

jsarha replied:
Probably, if we want to be sure the buffer size is longer than 2ms. It would also make sense to move the "2 *" inside the bracket to have the maximum precision in for the division.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1215,13 +1250,39 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			 * Despite its name the bitfield 'fifo_size' is used to define DMA buffer
+			 * size. The expression calculates 2ms buffer size.
+			 */
+			fifo_size = 2 * ((params_rate(fe_params) * params_channels(fe_params) *
```
should the 2 be a #define?

jsarha replied:
Would not hurt, but where should I put it? Or do we have such a macro already.

jsarha replied:
Well, I put it in the beginning of the ipc4-topology.c for the time being.

plbossart commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -72,6 +72,7 @@
  * @lp_mode: Low power mode
  * @mem_usage: Memory usage
  * @state: Pipeline state
+ * @use_chain_dma: flag to indicate if the firmware should use chained DMA
```
shall use chained DMA. It's not a recommendation, it's a requirement. The topology would not work if the firmware ignored this information.

plbossart reviewed (commented):
minor nit-picks @jsarha

jsarha commented:
Addressed @plbossart 's comments and changed -ENOTSUPP to -ENODEV due to checkpatch error:

WARNING: ENOTSUPP is not a SUSV4 error code, prefer EOPNOTSUPP
#313: FILE: sound/soc/sof/ipc4-topology.c:528:
+               ret = -ENOTSUPP;

plbossart commented:
> Addressed @plbossart 's comments and changed -ENOTSUPP to -ENODEV due to checkpatch error:
> 
> WARNING: ENOTSUPP is not a SUSV4 error code, prefer EOPNOTSUPP #313: FILE: sound/soc/sof/ipc4-topology.c:528:
> 
> * ```
>             ret = -ENOTSUPP;
>   ```

oh, that one you can ignore @jsarha. There is no consensus in maintainer circles on this checkpatch warning so we keep using ENOTSUPP

if it is confusing it's because it is.

jsarha commented:
> oh, that one you can ignore @jsarha. There is no consensus in maintainer circles on this checkpatch warning so we keep using ENOTSUPP
> 
> if it is confusing it's because it is.

I do not think there is any pressing need to use ENOTSUPP either, should I still revert back to ENOTSUPP from ENODEV?

jsarha commented:
I have also rebased this PR on top of #3972 if it is merged first.

jsarha commented:
@plbossart here is the version that is rebased on top of the latest #3972 :
https://github.com/jsarha/linux-sof/pull/new/chain_dma-on_top_of_3972

The only place that looks different to current version of this PR is hda-dai.c and hda-dai-ops.c. I also tested with the rebased PR that playback survives susped-resume cycle when playing and when paused.

jsarha commented:
Now the PR is rebased on top of #3972 , so please ignore everything else but the three last commits.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,69 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+
+	/*
+	 * To set-up the DMA chain, the host DMA ID and SCS setting
+	 * are retrieved fromthe host pipeline configuration. Likewise
```
looks great @jsarha. Just a tiny typo here "fromthe"

plbossart replied:
typo still there @jsarha

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -216,6 +281,21 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+
+		/* If a pipeline in the list uses chained DMA, trigger chained DMA and return */
+		if (pipeline->use_chain_dma)
+			return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state);
```
I find this loop a bit confusing.

The comments talk about "every pipeline of a chained DMA" and "if a pipeline uses chained DMA", but in practice the code triggers the FIRST pipeline that's found.

We should clarify that for chain DMA this is intentional, the pipelines are notional and using the first is just fine.

jsarha replied:
Added more comments to this place and before sof_ipc4_chain_dma_trigger().

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -368,7 +448,11 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 	struct snd_interval *rate = hw_param_interval(params, SNDRV_PCM_HW_PARAM_RATE);
 	struct snd_mask *fmt = hw_param_mask(params, SNDRV_PCM_HW_PARAM_FORMAT);
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_dai *cpu_dai = asoc_rtd_to_cpu(rtd, 0);
```
assigned but not used?

jsarha replied:
This was actually wrong fix, cpu_dai is used inside the loop.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -199,6 +262,8 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
 	struct ipc4_pipeline_set_state_data *trigger_list;
+	struct snd_sof_widget *pipe_widget;
+	struct sof_ipc4_pipeline *pipeline;
```
move to scope of for loop starting line 289?

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -368,7 +448,11 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_dai *cpu_dai = asoc_rtd_to_cpu(rtd, 0);
 	struct sof_ipc4_copier *ipc4_copier;
+	struct snd_soc_dapm_widget *w;
```
move to scope of for loop starting line 471?

plbossart reviewed (commented):
Look quite good indeed @jsarha, couple of minor comments below. Nice work!

jsarha commented:
@plbossart 's comments addressed.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,78 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+ * Chained DMA is a special case where there is no processing on
+ * DSP. The samples are just moved over host side DMA to a single
+ * buffer on DSP and directly from the to link DMA.  However, the
+ * model on SOF driver has two dummy pipeline, one at host DAI, and
```
while you're still fixing typos, here's another one "two dummy pipelines" and we've been asked to use inclusive terms. so maybe two pretend pipelines?

plbossart replied:
I love using "notional", it makes me sound smart.

jsarha replied:
/me checks exact meaning of notional

Seems perfect, and I learned a new word in English :). And fixed one other typo that I found after reading the text for the second time.

jsarha commented:
Some typo and wording fixes applied, end removed one bad fix. Now it works again 😅.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -383,9 +479,25 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
-	snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+	for_each_pcm_streams(dir) {
+		struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, dir);
+		if (w) {
```
doesnt checkpatch complain about a missing empty line after declarations?

plbossart commented:
https://github.com/thesofproject/linux/pull/3972 is merged, please rebase @jsarha. Should be straightforward.

jsarha commented:
Rebased on top of the latest sof-dev, fixed couple of typos more, and the checkpatch issue I added last night. Should be good now, please approve @ujfalusi , @ranj063 , @plbossart !

RanderWang commented on sound/soc/sof/intel/hda-dai-ops.c:
```diff
@@ -330,9 +339,18 @@ hda_select_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *swidg
+			if (pipeline->use_chain_dma)
+				return &hda_ipc4_chain_dma_ops;
+			else
+				return &hda_ipc4_dma_ops;
```
change to 
````c
if (pipeline->use_chain_dma)
     return &hda_ipc4_chain_dma_ops;
			
return &hda_ipc4_dma_ops;
````

RanderWang reviewed (approved):
LGTM

jsarha commented:
Addressed @RanderWang 's comment and while at it moved the pipe_widget and pipeline variables down in scope.

RanderWang reviewed (approved):
LGTM

plbossart reviewed (approved):
LGTM, thanks @jsarha

ranj063 commented:
@plbossart @jsarha please don't merge this PR. HDA device with chained DMA has a lot of failures

ranj063 commented:
@jsarha the fixes for SOF firmware failures are in https://github.com/thesofproject/linux/pull/4030

jsarha commented:
The problem was SOF FW being sensitive about fifo_size field being present in the message. I developed the whole time against cavs and forgot to test against SOF too. The problem should now be fixed. The changes lie in sof_ipc4_chain_dma_trigger() and in sof_ipc4_chain_dma_send_msg(). @ujfalusi , @ranj063 , @plbossart , and @RanderWang , please review again.

ujfalusi commented:
> The problem was SOF FW being sensitive about fifo_size field being present in the message. I developed the whole time against cavs and forgot to test against SOF too. The problem should now be fixed. The changes lie in sof_ipc4_chain_dma_trigger() and in sof_ipc4_chain_dma_send_msg(). @ujfalusi , @ranj063 , @plbossart , and @RanderWang , please review again.

It must not matter if it is cavs or SOF.ipc4 firmware, both should work in a same way, no?

jsarha commented:
> > The problem was SOF FW being sensitive about fifo_size field being present in the message. I developed the whole time against cavs and forgot to test against SOF too. The problem should now be fixed. The changes lie in sof_ipc4_chain_dma_trigger() and in sof_ipc4_chain_dma_send_msg(). @ujfalusi , @ranj063 , @plbossart , and @RanderWang , please review again.
> 
> It must not matter if it is cavs or SOF.ipc4 firmware, both should work in a same way, no?

It appears that they do not, but its possible to write a driver that works with both. cavs probably interprets the messages based on the stream state, but SOF tries to decode the state from the message, and ends up doing wrong things if fifo_size is there when its not needed.

jsarha commented:
There is still something fishy when running multiple-pause-resume test with SOF FW. cavs works, but SOF does not. Trying to debug what is going on.

ranj063 commented:
SOFCI TEST

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2165,25 +2165,19 @@ static int sof_complete(struct snd_soc_component *scomp)
-	list_for_each_entry(swidget, &sdev->widget_list, list) {
-		if (widget_ops && widget_ops[swidget->id].ipc_setup) {
-			ret = widget_ops[swidget->id].ipc_setup(swidget);
+	/* set the pipe_widget and apply the dynamic_pipeline_widget_flag */
```
This comment does not align with the code:
The pipe_widget should be plural (you are going through the pipeline list)?
What is 'apply(ing) the dynamic_widget_flag'?

Down somewhere it might be applying the flag, but it is just "set up the IPC structures for the pipeline widgets"

jsarha replied:
This patch is originally from @ranj063 and it was introduced (IIRR) to have pipeline->use_chain_dma set already at sof_ipc4_pcm_dai_link_fixup() call time.  I have rebased this commint a couple of times and may have mangled the comments.  "set up the IPC structures for the pipeline widgets" sounds good to em, but let's hear the author before changing it.

ranj063 replied:
yes looks like a stale comment. lets fix it

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2165,25 +2165,19 @@ static int sof_complete(struct snd_soc_component *scomp)
+		struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+
+		/* Update the scheduler widget's IPC structure */
+		if (widget_ops[pipe_widget->id].ipc_setup) {
```
Well, there is no indication that this is filtering for 'scheduler widget'. We go through the pipelines and update the pipe_widget associated with each of the pipeline. if the top level comment is correct this is not needed (and it is confusing in current form).
Yes, the pipe_widget _might_ be a scheduler widget, but without intimate knowledge or looking up in code it is not obvious.

ranj063 replied:
I suppose we could move this inside the same list iteration on line 2187

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2199,6 +2193,22 @@ static int sof_complete(struct snd_soc_component *scomp)
+	 */
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (swidget->id != snd_soc_dapm_scheduler && widget_ops[swidget->id].ipc_setup) {
+			ret = widget_ops[swidget->id].ipc_setup(swidget);
```
can this be done right after after `sof_set_widget_pipeline()` call above or this needs to be done for widgets not part of any pipeline? DO we even have such a widget?

just too many `list_for_each_entry(swidget, &sdev->widget_list, list)` in sequence.

It might result too deep indentation?

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -216,6 +296,27 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
+		 */
+		if (pipeline->use_chain_dma)
+			return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);
+	}
```
I guess there is no way to reduce the need for the loop by checking first the `[0]->pipe_widget->use_chain_dma` as it is plausible that some of the pipelines in the chain might have the flag set, making the path unworkable?

jsarha replied:
Like the comments indicate, the check you suggest would be enough for the functionality. However, this loop is there to verify the use_chain_dma attributes are used consistently in the topology. e.g. handle every pipeline_list with even a single use_chain_dma attribute with sof_ipc4_chain_dma_trigger(), which in turn makes sure all the pipelines in the list have the attribute, and produces an appropriate error print if they don't.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -631,6 +646,14 @@ static int sof_ipc4_widget_setup_comp_pipeline(struct snd_sof_widget *swidget)
+
+		swidget->private = pipeline;
+
+		return 0;
```
do you need these blank lines?

jsarha replied:
Nope.

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1215,13 +1250,41 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			u32 fifo_size;
+
+			host_dma_id = platform_params->stream_tag - 1;
+			pipeline->msg.primary |= SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(host_dma_id);
```
do you need to clear the bits before setting it? Can we have a case when the pipeline is use on different DMA channel?

jsarha replied:
The bits are cleared in sof_ipc4_unprepare_copier_module() [1]. I used to clear them also here where the assignment is done, but in some earlier comment I was told not to.

[1] https://github.com/jsarha/linux-sof/blob/8b1f2414aac5a4d1678ea7de6373a3929604efb0/sound/soc/sof/ipc4-topology.c#L1079

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1215,13 +1250,41 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+						  params_rate(fe_params) *
+						  params_channels(fe_params) *
+						  params_physical_width(fe_params)), 8000);
+			pipeline->msg.extension |= SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE(fifo_size);
```
this might need a clear before setting? The FIFO size depends on channels and sample size...

jsarha replied:
It is also cleared in sof_ipc4_unprepare_copier_module().

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2002,12 +2089,20 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
 
+	/* no route set up if chain DMA is used */
+	if (src_pipeline->use_chain_dma || sink_pipeline->use_chain_dma)
+		return 0;
```
question: in reality we will have two pipelines and both have the use_chain_dma set, right? And afaik it is an error if any side does not have this flag, we would not be here. I think.
If we have a path which should use ChainDMA then _all_ pipelines in it must have the flag, so it is enough to check in the src, the test on the sink will never happen, right?

jsarha replied:
Yes its an error if only one of the pipelines has use_chain_dma attribute, but I think its better to avoid misconfiguration at this phase and produce a clear error at trigger phase, about the problem. See my reply to your earlier comment here: https://github.com/thesofproject/linux/pull/3953#discussion_r1033267616

ujfalusi replied:
Right, what I meant is that if we are here then we have already went through the sanity check of the pipeline path that all (or none of them) have the flag set, thus if any of the pipelines in the path have the flag then for sure all will have it, no?

jsarha replied:
Have we? The order the each callback is called is not completely clear to me, but I would have expected route_setup to be before trigger. However, if its not then sure I'll remove one of the checks.

ujfalusi replied:
hrm, so you do double sanity check?
You are right that the route setup should be prior to trigger, so in trigger you should already have a route which either uses ChainDMA or  not use ChainDMA as you already failed the erroneous mixed use?

jsarha replied:
The check is in trigger, because there I will anyway have to pull everything together process both pipelines in the relevant pipeline_list, but yes the check could be here too.  Should I move it here, and simplify the trigger function a little? Or is it still better that as much as possible of the chain dma code is in one place in trigger function?

jsarha replied:
Changed it after all to have the check in sof_ipc4_route_setup() and to simplify the sof_ipc4_trigger_pipelines() a bit.

ujfalusi reviewed (commented):
Tests are passing, so this looks good, I have some questions..

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,86 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+		if (cmd == SNDRV_PCM_TRIGGER_PAUSE_RELEASE)
+			return sof_ipc4_chain_dma_send_msg(sdev, pipeline_list, true, true, false);
+		else
+			return sof_ipc4_chain_dma_send_msg(sdev, pipeline_list, true, true, true);
```
You don't need the `else`

ujfalusi commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -191,6 +191,86 @@ sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+	default:
+		dev_err(sdev->dev, "Unexpected state %d", state);
+		return -EINVAL;
+	}
```
Would it make sense to have the bools in this function, set them and move the bulk of the `sof_ipc4_chain_dma_send_msg()` after the switch() ?

jsarha replied:
I think it would.

ranj063 replied:
@jsarha I think @ujfalusi was suggesting to merge sof_ipc4_chain_dma_send_msg() with this function too.

jsarha replied:
Oh yes. When I got half of the idea, was too busy to implement it, to understand the other half.

jsarha commented:
Yet one more, hopefully last, round. @ujfalusi  , @ranj063 , @plbossart , and @RanderWang , please review again.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1053,11 +1074,23 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
 		ipc4_copier = dai->private;
+
+		if (pipeline->use_chain_dma) {
+			/* Zero DAI end pipeline msg contents */
```
I think this comment and the one above are not necessary, The code is pretty self-explanatory in these cases

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1215,13 +1248,41 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
+			pipeline->msg.extension |= SOF_IPC4_GLB_EXT_CHAIN_DMA_FIFO_SIZE(fifo_size);
+
+			/* Set node_id to invalid value. It shouldn't be used with chained DMA. */
+			copier_data->gtw_cfg.node_id = SOF_IPC4_INVALID_NODE_ID;
```
why is this needed? Its not like we ever send the IPC to set up the copier widget for chained DMA, right?

jsarha replied:
IIRR it was @RanderWang 's suggestion. AFAIU the reasoning is to cause an error, if a chain dma swidget is ever used in a wrong place.

ranj063 replied:
@jsarha that would be wrong and would lead to memory leaks. We allocate memory based on the node ID durig setup() and if you modify it during prepare, what happens during free would be anybody's guess.

jsarha replied:
Wrong yes, the idea is only to catch such a bug early. Or do you mean that just making the assignment here would cause a leak? That would be surprising as  copier_data->gtw_cfg.node_id is assigned in so many other branches of the function.

ranj063 replied:
sorry I got confused. I thpought we allocated memory based on node_id but its really only based on dai_type. So I guess this is fine

ranj063 replied:
But I still think it is not right set to invalid here. The idea is to keep the tplg identical for both chained DMA and regular pipelines. So I do not see what error you would catch with this.

jsarha replied:
Here is the original comment from @RanderWang 's: https://github.com/thesofproject/linux/pull/3953#discussion_r1017558014 

I guess we can wait until he wakes up, if there is something more to the suggestion than I  can explain, before making the final decision.

RanderWang replied:
Chain-DMA does't support time stamp feature which will check node to skip this case.

RanderWang replied:
````c
static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
					    struct snd_pcm_substream *substream,
					    struct snd_sof_pcm_stream *stream,
					    struct sof_ipc4_timestamp_info *time_info)
{
	struct sof_ipc4_copier *host_copier = time_info->host_copier;
	struct sof_ipc4_copier *dai_copier = time_info->dai_copier;
	struct sof_ipc4_pipeline_registers ppl_reg;
	u64 stream_start_position;
	u32 dai_sample_size;
	u32 ch, node_index;
	u32 offset;

	node_index = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
	if (node_index == SOF_IPC4_INVALID_NODE_ID)
		return -EINVAL;
````

ranj063 replied:
@jsarha so it looks like we need to set the node ID to invalid for this particular reason that we do no support timestamping feature with chained DMA. Can you please update your comment here so it is clear why we need this?

RanderWang commented on sound/soc/sof/topology.c:
```diff
@@ -2165,37 +2164,37 @@ static int sof_complete(struct snd_soc_component *scomp)
+				ret = sof_set_widget_pipeline(sdev, spipe, swidget);
 				if (ret < 0)
 					return ret;
+				if (widget_ops[swidget->id].ipc_setup) {
```
add white space line before "if (widget_ops[swidget->id].ipc_setup) {"

jsarha commented:
Hmm. Just setting node_id to invalid does not appear to be enough:

Nov 30 09:09:17 jhary kernel: divide error: 0000 [#1] PREEMPT SMP NOPTI
Nov 30 09:09:17 jhary kernel: CPU: 0 PID: 1594 Comm: aplay Not tainted 6.1.0-rc6test-84170-g300fc52a0bc0 #246
Nov 30 09:09:17 jhary kernel: Hardware name: AAEON UPX-TGL01/UPX-TGL01, BIOS UXTGBM14 11/05/2021
Nov 30 09:09:17 jhary kernel: RIP: 0010:sof_ipc4_pcm_delay+0x192/0x1b0 [snd_sof]
Nov 30 09:09:17 jhary kernel: Code: 24 90 03 00 00 e8 6e 1b ff ff 48 8b 44 24 04 48 83 f8 ff 0f 84 5a ff ff ff 41 8b 4e 2c 41 0f b6 56 3c c1 e9 03 0f af ca 31 d2>
Nov 30 09:09:17 jhary kernel: RSP: 0018:ffffad030303bc10 EFLAGS: 00010246
Nov 30 09:09:17 jhary kernel: RAX: 2e79646165727766 RBX: ffff9d76c5da3000 RCX: 0000000000000000
Nov 30 09:09:17 jhary kernel: RDX: 0000000000000000 RSI: ffffad0301781040 RDI: ffffad030303bc24
Nov 30 09:09:17 jhary kernel: RBP: ffff9d76c707ca28 R08: ffffad030303bc14 R09: 0000000000000000
Nov 30 09:09:17 jhary kernel: R10: ffffad030303bd70 R11: 0000000000000001 R12: ffff9d76c35ce028
Nov 30 09:09:17 jhary kernel: R13: ffff9d76c5ee2a80 R14: ffff9d76c5cd2f00 R15: ffff9d76c5205028
Nov 30 09:09:17 jhary kernel: FS:  00007efd539c8440(0000) GS:ffff9d7843800000(0000) knlGS:0000000000000000
Nov 30 09:09:17 jhary kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Nov 30 09:09:17 jhary kernel: CR2: 00007efd53e74008 CR3: 000000013569e001 CR4: 0000000000770ef0
Nov 30 09:09:17 jhary kernel: PKRU: 55555554
Nov 30 09:09:17 jhary kernel: Call Trace:
Nov 30 09:09:17 jhary kernel:  <TASK>
Nov 30 09:09:17 jhary kernel:  snd_soc_pcm_component_delay+0x56/0xd0 [snd_soc_core]
Nov 30 09:09:17 jhary kernel:  soc_pcm_pointer+0x5f/0xa0 [snd_soc_core]
Nov 30 09:09:17 jhary kernel:  snd_pcm_update_hw_ptr0+0x4c/0x870 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  ? __snd_pcm_lib_xfer+0xcc/0x910 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  ? find_held_lock+0x32/0x90
Nov 30 09:09:17 jhary kernel:  ? snd_pcm_hw_constraint_mask64+0x70/0x70 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  __snd_pcm_lib_xfer+0x84d/0x910 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  ? snd_pcm_hw_rule_step+0x100/0x100 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  ? _raw_spin_unlock_irqrestore+0x33/0x60
Nov 30 09:09:17 jhary kernel:  ? __wake_up_common_lock+0x8a/0xc0
Nov 30 09:09:17 jhary kernel:  snd_pcm_common_ioctl+0xca7/0x11c0 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  ? ioctl_has_perm.constprop.0+0xdb/0x140
Nov 30 09:09:17 jhary kernel:  snd_pcm_ioctl+0x23/0x40 [snd_pcm]
Nov 30 09:09:17 jhary kernel:  __x64_sys_ioctl+0x88/0xc0
Nov 30 09:09:17 jhary kernel:  do_syscall_64+0x38/0x90
Nov 30 09:09:17 jhary kernel:  entry_SYSCALL_64_after_hwframe+0x63/0xcd
Nov 30 09:09:17 jhary kernel: RIP: 0033:0x7efd53c323db
Nov 30 09:09:17 jhary kernel: Code: 0f 1e fa 48 8b 05 b5 7a 0d 00 64 c7 00 26 00 00 00 48 c7 c0 ff ff ff ff c3 66 0f 1f 44 00 00 f3 0f 1e fa b8 10 00 00 00 0f 05>
Nov 30 09:09:17 jhary kernel: RSP: 002b:00007ffe31a9b268 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
Nov 30 09:09:17 jhary kernel: RAX: ffffffffffffffda RBX: 000055f6040e7770 RCX: 00007efd53c323db
Nov 30 09:09:17 jhary kernel: RDX: 00007ffe31a9b270 RSI: 0000000040184150 RDI: 0000000000000004
Nov 30 09:09:17 jhary kernel: RBP: 000055f6040e76f0 R08: 0000000000000000 R09: 000055f6040e9720
Nov 30 09:09:17 jhary kernel: R10: 000055f6040c5010 R11: 0000000000000246 R12: 000055f6040e96d0
Nov 30 09:09:17 jhary kernel: R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000800
Nov 30 09:09:17 jhary kernel:  </TASK>
Nov 30 09:09:17 jhary kernel: Modules linked in: fuse snd_soc_skl_hda_dsp snd_soc_intel_hda_dsp_common snd_soc_hdac_hdmi snd_sof_ipc_msg_injector snd_sof_probes >
Nov 30 09:09:17 jhary kernel:  e1000e xhci_pci mfd_core drm drm_panel_orientation_quirks xhci_hcd video wmi
Nov 30 09:09:17 jhary kernel: ---[ end trace 0000000000000000 ]---
Nov 30 09:09:17 jhary kernel: RIP: 0010:sof_ipc4_pcm_delay+0x192/0x1b0 [snd_sof]
Nov 30 09:09:17 jhary kernel: Code: 24 90 03 00 00 e8 6e 1b ff ff 48 8b 44 24 04 48 83 f8 ff 0f 84 5a ff ff ff 41 8b 4e 2c 41 0f b6 56 3c c1 e9 03 0f af ca 31 d2>
Nov 30 09:09:17 jhary kernel: RSP: 0018:ffffad030303bc10 EFLAGS: 00010246
Nov 30 09:09:17 jhary kernel: RAX: 2e79646165727766 RBX: ffff9d76c5da3000 RCX: 0000000000000000
Nov 30 09:09:17 jhary kernel: RDX: 0000000000000000 RSI: ffffad0301781040 RDI: ffffad030303bc24
Nov 30 09:09:17 jhary kernel: RBP: ffff9d76c707ca28 R08: ffffad030303bc14 R09: 0000000000000000
Nov 30 09:09:17 jhary kernel: R10: ffffad030303bd70 R11: 0000000000000001 R12: ffff9d76c35ce028
Nov 30 09:09:17 jhary kernel: R13: ffff9d76c5ee2a80 R14: ffff9d76c5cd2f00 R15: ffff9d76c5205028
Nov 30 09:09:17 jhary kernel: FS:  00007efd539c8440(0000) GS:ffff9d7843800000(0000) knlGS:0000000000000000
Nov 30 09:09:17 jhary kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Nov 30 09:09:17 jhary kernel: CR2: 00007efd53e74008 CR3: 000000013569e001 CR4: 0000000000770ef0
Nov 30 09:09:17 jhary kernel: PKRU: 55555554

jsarha commented:
I do not see a single division in snd_soc_pcm_component_delay(), so where does the "divide error" come from?
... more coffee.

jsarha commented:
... one more time. @ujfalusi , @ranj063 , @plbossart , and @RanderWang , please review again.

RanderWang commented:
> Hmm. Just setting node_id to invalid does not appear to be enough:
> 
> Nov 30 09:09:17 jhary kernel: divide error: 0000 [#1] PREEMPT SMP NOPTI Nov 30 09:09:17 jhary kernel: CPU: 0 PID: 1594 [snd_sof] Nov 30 09:09:17 jhary kernel: Code: 24 90 03 00 00 e8 6e 1b ff ff 48 8b 44 24 04 48 83 f8 ff 0f 84 5a ff ff ff 41 8b 4e 2c  R12: ffff9d76c35ce028 Nov 30 09:09:17 jhary kernel: R13: ffff9d76c5ee2a80 R14: ffff9d76c5cd2f00 R15: ffff9d76c5205028 Nov 30 09:09:17 jhary kernel: FS: 00007efd539c8440(0000) GS:ffff9d7843800000(0000) knlGS:0000000000000000 Nov 30 09:09:17 jhary kernel: CS: 0010 DS: 0000 ES: 0000 CR0: 0000000080050033 Nov 30 09:09:17 jhary kernel: CR2: 00007efd53e74008 CR3: 000000013569e001 CR4: 0000000000770ef0 Nov 30 09:09:17 jhary kernel: PKRU: 55555554

@jsarha  can you try https://github.com/thesofproject/linux/pull/4050. Thanks!

RanderWang commented:
@jsarha  please skip the first commit since there is a fixup, thanks

jsarha commented:
> @jsarha please skip the first commit since there is a fixup, thanks

Sure, I'll rebase as soon as #4050 is merged, and drop the first patch.

ujfalusi reviewed (approved):
@jsarha, looks good, one nitpcik of mine is the setting of bools in one line in `sof_ipc4_chain_dma_trigger()`, but nothing else.

jsarha commented:
Now that #4050 is merged, there should nothing stopping from approving and merging this too (but I have been wrong many times before 😅) . @ujfalusi , @ranj063 , @plbossart , and @RanderWang , please review again.

ranj063 commented:
> Now that #4050 is merged, there should nothing stopping from approving and merging this too (but I have been wrong many times before sweat_smile) . @ujfalusi , @ranj063 , @plbossart , and @RanderWang , please review again.

Nice work and thanks for being patient with this one @jsarha.

ranj063 commented:
SOFCI TEST

ujfalusi reviewed (approved):
@jsarha, thank you for the update, no more nitpicks from my side.
